### PR TITLE
NuttX enable SPI DMA per board

### PR DESCRIPTION
**Pull request by dagar**
- [ ] airmind/mindpx-v2
 - [x] ~~auav/esc35-v1~~
 - [x] auav/x21
 - [x] av/x-v1
 - [x] ~~bitcraze/crazyflie~~
 - [ ] holybro/kakutef7
 - [ ] intel/aerofc-v1
 - [x] ~~nxp/fmuk66-v3~~
 - [ ] omnibus/f4sd
 - [x] ~~px4/cannode-v1~~
 - [x] ~~px4/esc-v1~~
 - [x] px4/fmu-v2
 - [x] px4/fmu-v3
 - [x] px4/fmu-v4
 - [x] px4/fmu-v4pro
 - [ ] px4/fmu-v5
 - [ ] px4/fmu-v5x
 - [x] ~~px4/io-v2~~
 - [x] ~~thiemar/s2740vc-v1~~

### CODE REVIEW

File: boards/aerotenna/ocpoc/src/CMakeLists.txt [hunk 1]
```diff
@@ -31,6 +31,6 @@
 #
 ############################################################################
 
-px4_add_library(drivers_board
+add_library(drivers_board
 	${PX4_SOURCE_DIR}/src/drivers/boards/common/board_identity.c # TODO: this is horrible and should be fixed
```
Comment (davids5) [inline boards/aerotenna/ocpoc/src/CMakeLists.txt]: what is horrible and what need to be fixed

Comment (dagar) [reply boards/aerotenna/ocpoc/src/CMakeLists.txt]: Cleanup along the way. Moved to https://github.com/PX4/Firmware/pull/12446.

**Comment by jorge789** [comment]
**Tested on PixRacer V4:**

**Modes Tested**
Position Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

**Note:**
No issues noted, good flight in general.

**Logs:**
https://review.px4.io/plot_app?log=3ca5d59f-e633-4281-ac5f-21974ae98f80

https://review.px4.io/plot_app?log=e48ab5c9-1a1d-4095-be23-23e55a800ef4

**Logs: Comparison to master**
https://review.px4.io/plot_app?log=29892ab0-32ee-40d8-a5fa-1636ab1f62df

https://review.px4.io/plot_app?log=b7466023-f8d2-4366-83b6-d4cef75e549f


**Tested on Pixhawk 2 Cube  V3:**

**Modes Tested**
Position Mode: Good.
Mission Plan Mode (Automated): Good.
RTL (Return To Land): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

**Note:**
No issues noted, good flight in general.

**Logs:**
https://review.px4.io/plot_app?log=4e7bffaf-0bb3-4b09-87e6-4304a56c2669

https://review.px4.io/plot_app?log=f467fd1d-c1ae-447a-ac57-815f05f02cfe

**Logs: Comparison to master**
https://review.px4.io/plot_app?log=1ee0bdfc-0a9e-4a47-8dec-593e8565d45a

**Comment by dannyfpv** [comment]
Tested on Pixhawk4 v5 f-450 
gps issue.
upload master firmware and no issues


Screen Shot 2019-07-09 at 9 51 25 AM
Screen Shot 2019-07-09 at 9 51 25 AM

**Comment by Junkim3DR** [comment]
### Tested on PixRacer V4:

**Modes Tested:**
Position Mode: Untested.
Altitude Mode: Untested.
Stabilized Mode: Untested
Mission Plan Mode (Automated): Untested
RTL: Untested

**Procedure:**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL and see landing behaviour.

**Notes:**
Vehicle did not have GPS Lock and was not able to be flown. 

**Log:**
- Master: https://review.px4.io/plot_app?log=20ddbd80-9dc5-4702-9ef2-6ac583c01450





### Tested on Pixhawk Pro v4:

**Modes Tested:**
- Position Mode: Good.
- Altitude Mode: Good.
- Stabilized Mode: Good.
- Mission Plan Mode (Automated): Good.
- RTL: Good.

**Procedure:**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL and see landing behaviour.

**Notes:**
No issues noted, good flight in general.

**Logs:**
- PR12436: https://review.px4.io/plot_app?log=a38d33f3-8023-437b-af09-a3d6df75d0bd

- Master: https://review.px4.io/plot_app?log=e67424c5-e435-49f1-9b82-c298f5e643d9





### Tested on NXP FMUK66 v3:

**Modes Tested:**
- Position Mode: Good.
- Altitude Mode: Good.
- Stabilized Mode: Good.
- Mission Plan Mode (Automated): Good.
- RTL: Good.

**Procedure:**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint activate RTL and see landing behaviour.

**Notes:**
No issues noted, good flight in general.

**Logs:**
- PR12436: https://review.px4.io/plot_app?log=3886a4e7-1c33-455b-998e-d9ab0db977e2

- Master: https://review.px4.io/plot_app?log=ab3333a4-175f-40f8-87d3-6dc6b927d0eb

**Comment by dagar** [comment]
Added holybro/kakutef7 to the list.

File: src/drivers/boards/common/board_dma_alloc.h [hunk 2]
```diff
@@ -112,7 +114,7 @@ __EXPORT int board_get_dma_usage(uint16_t *dma_total, uint16_t *dma_used, uint16
 #if defined(BOARD_DMA_ALLOC_POOL_SIZE)
```
Comment (dakejahl) [inline src/drivers/boards/common/board_dma_alloc.h]: @dagar Pardon my noobness, I want to make sure I understand what's going on. So the way I understand this is: by providing this definition *BOARD_DMA_ALLOC_POOL_SIZE* we are telling the linker about the `board_dma_*` functions (that is what the __EXPORT does) which makes the functions available for use as long as we `#include <drivers/boards/common/board_dma_alloc.h>`?

Comment (dakejahl) [reply src/drivers/boards/common/board_dma_alloc.h]: Also, is `board_dma_alloc.cpp` something you wrote?

Comment (dagar) [reply src/drivers/boards/common/board_dma_alloc.h]: Yes, and it's just a slight generalization of the existing FAT DMA allocator.

**Comment by dagar** [comment]
This is currently a problem on stm32f7 with the DCACHE and DMA capable check where the SPI transfer buffer is being incorrectly considered non-DMA capable if it's not sized as a cache line multiple.

**Comment by dakejahl** [comment]
Why don't we just add a check to ensure we always allocate the right amount + padding to get us to the multiple of 4 bytes

**Comment by dagar** [comment]
> Why don't we just add a check to ensure we always allocate the right amount + padding to get us to the multiple of 4 bytes
> 
> The SPI transfer can still just use the actual size of the data it wants. i.e
> 

The problem is the DCACHE alignment check for the buffer end within NuttX.

On stm32f7 spi_exchange() is checking if the buffer is DMA capable.
https://github.com/PX4/NuttX/blob/d8da511082646d83a54c6905daca13f0a1a609f0/arch/arm/src/stm32f7/stm32_spi.c#L1681-L1682

The DMA capable check includes the buffer end.
https://github.com/PX4/NuttX/blob/d8da511082646d83a54c6905daca13f0a1a609f0/arch/arm/src/stm32f7/stm32_dma.c#L928

**Comment by dakejahl** [comment]
So all DMA spi transfers must be multiples of 4 bytes? That seems wrong.

**Comment by dagar** [comment]
> So all DMA spi transfers must be multiples of 4 bytes? That seems wrong.

Yes, the check in its current form is bogus.
1. The memory must be DMA capable (not CCM on F4 or DTCM on F7)
2. On F7 with DCACHE the buffer must be DCACHE aligned (32 bytes) both start and end (we now have writethrough disabled).


One idea David and I have discussed is to change the notion of "DMA capable" within NuttX, possibly similar to the DMA allocator we have in PX4. Then instead of the DMA capable checking memory range, etc it would be based on originating from this particular allocator.

**Comment by julianoes** [comment]
I suppose I should leave these branches around :man_shrugging:.

**Pull request closed without merging**